Visitor.NET.AutoVisitableGen/Visitor.NET.AutoVisitableGen/AutoVisitableIncrementalSourceGenerator.cs

```diff
@@ -105,28 +119,54 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
 
             var namespaceName = classSymbol.ContainingNamespace.ToDisplayString();
 
-            var code = $@"// <auto-generated/>
-
-using Visitor.NET;
-
-namespace {namespaceName};
-
-public partial {typeKind.ToString().ToLower()} {visitableTypeName} :
-    IVisitable<{visitableTypeName}>
-{{
-    public override TReturn Accept<TReturn>(
-        IVisitor<{baseTypeName}, TReturn> visitor) =>
-        Accept(visitor);
-
-    public TReturn Accept<TReturn>(
-        IVisitor<{visitableTypeName}, TReturn> visitor) =>
-        visitor.Visit(this);
-}}
-";
+            List<INamedTypeSymbol> containingTypes = SymbolsHelper.GetContainingTypes(classSymbol);
```

Comment:
It seems to me that we can collect all the necessary information for generation in one pass in GetTypeDeclarationForSourceGen method

Another type can also be nested in an interface or in a static class and we need to be careful with type modifiers 

here is draft sample how we can collect information about outer declaration's 

`    private static List<OuterItemInfo>? GetOuterItemInfo(TypeDeclarationSyntax typeDeclarationSyntax)
    {
        // todo we should have a look
        // at https://andrewlock.net/creating-a-source-generator-part-9-avoiding-performance-pitfalls-in-incremental-generators/#4-watch-out-for-collection-types
       
        var outer = new List<TypeDeclarationSyntax>(); 
        while (typeDeclarationSyntax.Parent is TypeDeclarationSyntax parent)
        {
            outer.Add(parent);
            typeDeclarationSyntax = parent;
        }

        if (!outer.Any())
        {
            // return empty list it's ok
            return [];
        }

        if (outer.TrueForAll(x => x.Modifiers.Any(SyntaxKind.PartialKeyword)))
        {
            return outer.Select(x =>
            {
                var kind = x switch
                {
                    ClassDeclarationSyntax => "class",
                    RecordDeclarationSyntax => "record",
                    InterfaceDeclarationSyntax => "interface"
                };
                var typeName = x.Identifier.Text;
                var modifiers = x.Modifiers;
                return new OuterItemInfo(kind, typeName, modifiers.ToFullString());
            }).ToList();
        }
        
        // it's mean that we cannot generate code for this type
        return null;
    }`

 FYI @YarinOmesi @Stepami

Reply:
@YarinOmesi nice observation!

@YarinOmesi Consider using AST traverse based algorithm linked above instead of semantic model access

Reply:
Using ast traverse I create list of `ContainingTypeInfo`  type I created